## <div style="text-align:center"><span style="color:red">PROJECT-2. Loading new data. Refinement of the analysis.</span></div>

### <div style="text-align:center"><span style="color:red">0. Setting up the environment for data processing.</span></div>

In [1]:
# loading ipython-sql magic extebtion:
%load_ext sql
# setting the initial rows display limitation:
%config SqlMagic.displaylimit=5
# setting the base connection variable:
conn = 'postgresql+psycopg2://****:****@****:****/****'

_<span style="color:green">To get the SQL requests working in this notebook ipython-sql magic is required. Use  
```$conda install -c conda-forge ipython-sql``` within conda environment or  
```$pip install ipython-sql``` for the pip case to install it.  
More info can be found on the [ipython-sql project repository](https://github.com/catherinedevlin/ipython-sql).  
The access to the database is also required._

### <div style="text-align:center"><span style="color:red">1. Displaying dataset tables to evaluate their size, structure, and content.</span></div>

<center> <img src = https://github.com/ssergeegress/SF-DS-Projects/blob/main/Project_01A/data/dst3-u2-pr2_1_1.png?raw=true alt='dst3-u2-pr2_1_1.png' style='width:800px'>

<div style="text-align:center"><span style="color:red">1.1 The 'Candidate' table from 'hh' database.</span></div>

In [2]:
%sql $conn SELECT * FROM hh.candidate ORDER BY id

44744 rows affected.


[('M', 39, 'Системный администратор', 390, 'частичная занятость, проектная работа, полная занятость', 'Системный администратор', datetime.date(2019, 4, 16), 1, Decimal('29000')),
 ('M', 60, 'Технический писатель', 838, 'частичная занятость, проектная работа, полная занятость', 'Менеджер проекта, Аналитик, Технический писатель', datetime.date(2019, 4, 12), 2, Decimal('40000')),
 ('F', 36, 'Оператор', 733, 'полная занятость', 'Кассир-операционист', datetime.date(2019, 4, 16), 3, Decimal('20000')),
 ('M', 38, 'Веб-разработчик (HTML / CSS / JS / PHP / базы данных; фреймворки, дизайн, интерфейсы, CMS)', 530, 'частичная занятость, проектная работа, полная занятость', 'Инженер-программист', datetime.date(2019, 4, 8), 4, Decimal('100000')),
 ('F', 26, 'Региональный менеджер по продажам', 802, 'полная занятость', 'Менеджер по продажам', datetime.date(2019, 4, 22), 5, Decimal('140000')),
 ('M', 29, 'Технический специалист', 530, 'полная занятость', 'Слесарь КИПиА', datetime.date(2019, 4, 22), 6, Decimal('25000')),
 ('M', 46, 'Руководитель ИТ-проектов', 530, 'полная занятость', 'Менеджер (Руководитель ИТ-проектов)', datetime.date(2019, 4, 25), 7, Decimal('250000')),
 ('M', 29, 'Инженер АСУ ТП', 530, 'полная занятость', 'Инженер АСУ ТП', datetime.date(2019, 5, 7), 8, Decimal('70000')),
 ('M', 29, 'Ревизор', 530, 'полная занятость', 'Старший специалист', datetime.date(2019, 4, 11), 9, Decimal('65000')),
 ('M', 34, 'Менеджер по работе с клиентами, Pre-sale менеджер, Менеджер по сопровождению клиентов', 530, 'полная занятость', 'Специалист отдела ИТ', datetime.date(2019, 4, 19), 10, Decimal('55000')),
 ('F', 35, 'Менеджер / оператор интернет-магазина', 530, 'полная занятость', 'Администратор интернет-магазина и салона', datetime.date(2019, 4, 11), 11, Decimal('55000')),
 ('M', 59, 'Технический писатель', 802, 'полная занятость', 'Ведущий специалист научно-технического отдела', datetime.date(2019, 4, 26), 12, Decimal('80000')),
 ('M', 39, 'Инженер-программист', 530, 'частичная занятость, полная занятость', 'Продавец-консультант', datetime.date(2019, 4, 26), 13, Decimal('40000')),
 ('F', 27, 'SMM-менеджер/ SEO/ creator', 323, 'полная занятость', 'SMM-менеджер/ SEO', datetime.date(2019, 4, 16), 14, Decimal('90000')),
 ('F', 21, 'Менеджер проекта', 979, 'полная занятость', 'Старший официант', datetime.date(2019, 4, 18), 15, Decimal('25000')),
 ('M', 31, 'Специалист', 530, 'частичная занятость, проектная работа, полная занятость', 'Контент-менеджер', datetime.date(2019, 4, 26), 16, Decimal('23000')),
 ('M', 29, 'Дизайнер продукта, UX/UI', 137, 'полная занятость', 'Дизайнер продукта, UX/UI дизайнер', datetime.date(2019, 4, 12), 17, Decimal('120000')),
 ('M', 37, 'Инженер Отдела IT', 530, 'частичная занятость', 'Дежурный IT-инженер', datetime.date(2019, 4, 11), 18, Decimal('30000')),
 ('M', 30, 'Ведущий программист 1C / Руководитель', 590, 'частичная занятость, полная занятость', 'Ведущий программист 1C', datetime.date(2019, 4, 26), 19, Decimal('260000')),
 ('M', 38, 'Менеджер интернет-магазина', 530, 'полная занятость', 'Руководитель отдела веб-продаж и сервиса', datetime.date(2019, 5, 7), 20, Decimal('55000')),
 ('M', 31, 'Сетевой инженер / Системный администратор', 607, 'частичная занятость, проектная работа, полная занятость', 'Системный администратор', datetime.date(2019, 4, 18), 21, Decimal('80000')),
 ('M', 27, 'Программист C++, Web-программист, разработчик аппаратуры,', 794, 'частичная занятость, полная занятость', 'Web-программист', datetime.date(2019, 4, 17), 22, Decimal('70000')),
 ('M', 36, 'Технический специалист', 645, 'полная занятость', 'Старший инженер по эксплуатации мобильной сети', datetime.date(2019, 4, 10), 23, Decimal('70000')),
 ('M', 33, 'Golang разработчик', 320, 'полная занятость', 'Программист Python', datetime.date(2019, 4, 25), 24, Decimal('60000')),
 ('M', 34, 'начальник отдела, главный специалист, аудитор, аналитик', 289, 'полная занятость', 'главный экономист - аналитик Отдела методологии и риск-технологий крупного 

<span style="color:red">1.2 The 'City' table from 'hh' database.</span>

In [3]:
%sql $conn SELECT * FROM hh.city

984 rows affected.


[('Абакан', 1),
 ('Абдулино', 2),
 ('Абинск', 3),
 ('Агой', 4),
 ('Агрыз', 5),
 ('Адлер', 6),
 ('Адыгейск', 7),
 ('Азнакаево', 8),
 ('Азов', 9),
 ('Аксай (Казахстан)', 10),
 ('Аксай (Ростовская область)', 11),
 ('Аксукент', 12),
 ('Актау', 13),
 ('Актобе', 14),
 ('Алабино', 15),
 ('Алапаевск', 16),
 ('Алга (Актюбинская обл)', 17),
 ('Алдан', 18),
 ('Алейск', 19),
 ('Александров', 20),
 ('Александровск (Пермский край)', 21),
 ('Александровск-Сахалинский', 22),
 ('Алексеевка', 23),
 ('Алексеевское', 24),
 ('Алексин', 25),
 ('Алматы', 26),
 ('Алушта', 27),
 ('Альметьевск', 28),
 ('Амурск', 29),
 ('Анапа', 30),
 ('Ангарск', 31),
 ('Анжеро-Судженск', 32),
 ('Апатиты', 33),
 ('Апрелевка', 34),
 ('Апшеронск', 35),
 ('Арамиль', 36),
 ('Ардатов', 37),
 ('Арзамас', 38),
 ('Армавир', 39),
 ('Армения', 40),
 ('Армянск', 41),
 ('Арсеньев', 42),
 ('Арсеньево', 43),
 ('Артем', 44),
 ('Архангельск', 45),
 ('Архипо-Осиповка', 46),
 ('Асбест', 47),
 ('Астана', 48),
 ('Астрахань', 49),
 ('Атырау', 50),
 ('Афипский', 51),
 ('Ахтубинск', 52),
 ('Ачинск', 53),
 ('Аша', 54),
 ('Бавлы', 55),
 ('Багдарин', 56),
 ('Байконур (Кызылорд. обл)', 57),
 ('Баймак', 58),
 ('Баку', 59),
 ('Балабаново', 60),
 ('Балаково', 61),
 ('Балахна', 62),
 ('Балашиха', 63),
 ('Балашов', 64),
 ('Балей', 65),
 ('Балтым', 66),
 ('Балхаш', 67),
 ('Барабинск', 68),
 ('Барановичи', 69),
 ('Барда (Пермский край)', 70),
 ('Барнаул', 71),
 ('Батайск', 72),
 ('Бахчисарай', 73),
 ('Бежецк', 74),
 ('Безенчук', 75),
 ('Бейнеу (Мангистауская обл)', 76),
 ('Белая Калитва', 77),
 ('Белая Холуница', 78),
 ('Белая Церковь', 79),
 ('Белгород', 80),
 ('Белебей', 81),
 ('Белев', 82),
 ('Белово', 83),
 ('Белогорск', 84),
 ('Белозерск', 85),
 ('Белоозёрский (Московская область)', 86),
 ('Белоомут', 87),
 ('Белорецк', 88),
 ('Белореченск', 89),
 ('Белоусово', 90),
 ('Белоярский', 91),
 ('Бердск', 92),
 ('Березники', 93),
 ('Березовский (Свердловская область)', 94),
 ('Бийск', 95),
 ('Биробиджан', 96),
 ('Бирск', 97),
 ('Бирюч', 98),
 ('Бишкек', 99),
 ('Благовещенск (Амурская область)', 100),
 ('Благовещенск (Республика Башкортостан)', 101),
 ('Благодарный', 102),
 ('Бобруйск', 103),
 ('Богородицк', 104),
 ('Богородск', 105),
 ('Бодайбо', 106),
 ('Болгария', 107),
 ('Бологое', 108),
 ('Большое Болдино', 109),
 ('Бор', 110),
 ('Борисов', 111),
 ('Борисовка', 112),
 ('Борисоглебск', 113),
 ('Боровичи', 114),
 ('Боровск', 115),
 ('Бородино', 116),
 ('Бразилия', 117),
 ('Братск', 118),
 ('Брест', 119),
 ('Бронницы', 120),
 ('Брюховецкая станица', 121),
 ('Брянск', 122),
 ('Бугры', 123),
 ('Бугульма', 124),
 ('Бугуруслан', 125),
 ('Буденновск', 126),
 ('Бузулук', 127),
 ('Буинск', 128),
 ('Бутурлино', 129),
 ('Бухара', 130),
 ('Буча', 131),
 ('Вагай', 132),
 ('Вад', 133),
 ('Валуйки', 134),
 ('Варениковская', 135),
 ('Васюринская', 136),
 ('Великие Луки', 137),
 ('Великий Новгород', 138),
 ('Великий Устюг', 139),
 ('Великобритания', 140),
 ('Вельск', 141),
 ('Венев', 142),
 ('Верхнеднепровский', 143),
 ('Верхняя Пышма', 144),
 ('Верхняя Салда', 145),
 ('Видное', 146),
 ('Витебск', 147),
 ('Вихоревка', 148),
 ('Владивосток', 149),
 ('Владикавказ', 150),
 ('Владимир', 151),
 ('Власиха', 152),
 ('Волгоград', 153),
 ('Волгодонск', 154),
 ('Волгореченск', 155),
 ('Волжск', 156),
 ('Волжский', 157),
 ('Вологда', 158),
 ('Волоколамск', 159),
 ('Вольск', 160),
 ('Воркута', 161),
 ('Воронеж', 162),
 ('Воскресенск', 163),
 ('Воткинск', 164),
 ('Всеволожск', 165),
 ('Вуктыл', 166),
 ('Выборг', 167),
 ('Выкса', 168),
 ('Выселки', 169),
 ('Высокая Гора', 170),
 ('Вышний Волочек', 171),
 ('Вьетнам', 172),
 ('Вяземский', 173),
 ('Вязники', 174),
 ('Вязьма', 175),
 ('Гай', 176),
 ('Ганцевичи', 177),
 ('Гатчина', 178),
 ('Гвардейск', 179),
 ('Геленджик', 180),
 ('Георгиевск', 181),
 ('Германия', 182),
 ('Глазов', 183),
 ('Глазуновка', 184),
 ('Глинищево', 185),
 ('Глуск', 186),
 ('Голицыно', 187),
 ('Гомель', 188),
 ('Горно-Алтайск', 189),
 ('Горнозаводск', 190),
 ('Горняк', 191),
 ('Городец', 192),
 ('Городище', 193),

<span style="color:red">1.3 The 'Candidate Timetable Type' table from 'hh' database.</span>

In [4]:
%sql $conn SELECT * FROM hh.candidate_timetable_type ORDER BY candidate_id DESC

31155 rows affected.


[(17732, 44744, 2),
 (12798, 44739, 2),
 (37394, 44738, 2),
 (37393, 44736, 2),
 (26767, 44735, 1),
 (31423, 44732, 1),
 (37392, 44731, 2),
 (26080, 44730, 1),
 (37391, 44728, 2),
 (14896, 44724, 1),
 (45408, 44723, 5),
 (64253, 44722, 1),
 (9899, 44721, 2),
 (75108, 44720, 2),
 (45404, 44719, 5),
 (31422, 44718, 2),
 (75106, 44717, 2),
 (14895, 44716, 2),
 (71020, 44715, 3),
 (71019, 44713, 2),
 (45403, 44712, 2),
 (12794, 44711, 1),
 (17731, 44709, 2),
 (45399, 44708, 1),
 (14892, 44707, 1),
 (39067, 44706, 5),
 (75105, 44705, 2),
 (37388, 44704, 1),
 (17730, 44703, 2),
 (71018, 44701, 2),
 (5270, 44700, 1),
 (45398, 44699, 2),
 (64251, 44698, 5),
 (26762, 44697, 1),
 (7293, 44696, 2),
 (23333, 44695, 2),
 (23332, 44694, 2),
 (71017, 44693, 2),
 (64250, 44690, 2),
 (37387, 44689, 2),
 (71015, 44688, 2),
 (37386, 44687, 5),
 (1345, 44686, 5),
 (71014, 44685, 2),
 (7292, 44684, 2),
 (12790, 44683, 5),
 (45396, 44682, 2),
 (23331, 44680, 2),
 (71013, 44679, 2),
 (7290, 44678, 1),
 (64249, 44675, 2),
 (64248, 44674, 2),
 (17728, 44672, 2),
 (17724, 44671, 1),
 (45393, 44670, 1),
 (31421, 44668, 2),
 (64243, 44667, 1),
 (4634, 44665, 1),
 (26079, 44658, 2),
 (37385, 44656, 2),
 (71012, 44654, 2),
 (31420, 44653, 2),
 (17872, 44652, 1),
 (64527, 44651, 2),
 (17723, 44650, 2),
 (45392, 44649, 2),
 (71011, 44648, 2),
 (8691, 44647, 2),
 (7289, 44645, 2),
 (17868, 44644, 2),
 (23328, 44642, 1),
 (26760, 44641, 2),
 (64242, 44640, 2),
 (71010, 44639, 3),
 (1704, 44638, 2),
 (71610, 44635, 1),
 (23323, 44633, 1),
 (45391, 44632, 1),
 (32567, 44630, 1),
 (9897, 44629, 2),
 (71006, 44627, 1),
 (23322, 44625, 2),
 (31415, 44624, 1),
 (23321, 44623, 2),
 (45387, 44622, 1),
 (37384, 44619, 2),
 (7285, 44618, 1),
 (23320, 44616, 2),
 (12789, 44615, 2),
 (31414, 44613, 1),
 (12788, 44612, 2),
 (75102, 44611, 2),
 (23450, 44610, 2),
 (39066, 44609, 2),
 (7281, 44608, 1),
 (71005, 44607, 2),
 (37383, 44606, 2),
 (14889, 44605, 1),
 (64241, 44604, 2),
 (71004, 44603, 2),
 (4632, 44602, 2),
 (37382, 44601, 2),
 (64240, 44600, 2),
 (9896, 44597, 2),
 (7277, 44595, 2),
 (45383, 44594, 1),
 (37381, 44592, 2),
 (7276, 44591, 2),
 (75100, 44590, 1),
 (37380, 44589, 2),
 (45378, 44587, 1),
 (14886, 44584, 2),
 (37378, 44583, 2),
 (9894, 44582, 1),
 (31413, 44580, 2),
 (4766, 44579, 2),
 (64238, 44578, 2),
 (45377, 44576, 2),
 (4631, 44574, 2),
 (8609, 44573, 2),
 (71003, 44571, 2),
 (23318, 44564, 2),
 (17722, 44563, 2),
 (71000, 44561, 1),
 (70998, 44560, 2),
 (26759, 44559, 2),
 (913, 44558, 2),
 (23317, 44557, 2),
 (2555, 44554, 4),
 (14885, 44552, 2),
 (4630, 44551, 2),
 (75294, 44550, 1),
 (23315, 44549, 1),
 (70997, 44548, 2),
 (31412, 44546, 2),
 (4629, 44545, 2),
 (2551, 44544, 1),
 (8689, 44542, 2),
 (17721, 44541, 2),
 (4628, 44540, 2),
 (45373, 44539, 1),
 (64235, 44538, 5),
 (31411, 44537, 2),
 (64234, 44536, 2),
 (37375, 44535, 1),
 (45369, 44531, 1),
 (12787, 44530, 5),
 (39063, 44529, 1),
 (31410, 44526, 2),
 (64230, 44525, 5),
 (37371, 44524, 5),
 (70993, 44523, 5),
 (12783, 44520, 1),
 (31409, 44519, 5),
 (45365, 44518, 1),
 (23314, 44517, 2),
 (31406, 44516, 1),
 (70992, 44515, 2),
 (75096, 44512, 5),
 (23313, 44511, 2),
 (70990, 44509, 2),
 (45361, 44508, 1),
 (75093, 44507, 2),
 (64306, 44506, 2),
 (70987, 44504, 1),
 (70986, 44503, 2),
 (18612, 44501, 5),
 (26078, 44497, 2),
 (31402, 44496, 5),
 (45360, 44495, 2),
 (70985, 44493, 5),
 (45359, 44492, 2),
 (70984, 44490, 2),
 (17719, 44489, 2),
 (45357, 44487, 2),
 (37370, 44485, 2),
 (23830, 44484, 2),
 (17716, 44483, 1),
 (17715, 44481, 2),
 (31398, 44480, 1),
 (75092, 44478, 2),
 (4627, 44477, 2),
 (5269, 44476, 2),
 (12782, 44475, 2),
 (23312, 44474, 2),
 (37368, 44473, 1),
 (75090, 44472, 2),
 (37365, 44471, 1),
 (70983, 44470, 2),
 (45356, 44469, 2),
 (45355, 44468, 2),
 (37364, 44465, 2),
 (31395, 44464, 1),
 (26077, 44463, 2),
 (12779, 44462, 1),
 (70982, 44459, 2),
 (12778, 44458, 2),
 (70981, 44457, 2),
 (9891, 44456, 1),
 (23310, 44455, 2),
 (8605, 44452, 1),
 (32566, 4445

<span style="color:red">1.4 The 'Timetable Type' table from 'hh' database.</span>

In [5]:
%sql $conn SELECT * FROM hh.timetable_type

5 rows affected.


[(1, 'гибкий график'),
 (2, 'полный день'),
 (3, 'сменный график'),
 (4, 'вахтовый метод'),
 (5, 'удаленная работа')]

<span style="color:red">1.5 Conclusion of a pivot table with the number of lines for clarity.</span>

In [6]:
%%sql $conn
/* just union the outputs of COUNT function */
SELECT
    'Candidate' table_name,
    COUNT(c.id) row_cnt
FROM
    hh.candidate c
UNION
SELECT
    'City',
    COUNT(ct.id) rows_cnt
FROM
    hh.city ct
UNION
SELECT
    'Candidate Timetable Type' table_name,
    COUNT(ctt.id) rows_cnt
FROM
    hh.candidate_timetable_type ctt
UNION
SELECT
    'Timetable Type',
    COUNT(tt.id)
FROM
    hh.timetable_type tt
ORDER BY 2 DESC



4 rows affected.


[('Candidate', 44744),
 ('Candidate Timetable Type', 31155),
 ('City', 984),
 ('Timetable Type', 5)]

- <span style="color:blue">Hmmm... Looks neat. Let's try to deal with this a little, guys!</span>

### <div style="text-align:center"><span style="color:red">2. Preliminary data analysis.</span></div>

<span style="color:red">2.1 Calculation of the maximum age (max_age) of the candidate in the table.</span>

In [7]:
%%sql $conn
/* using MAX function to get the maximum val */ 
SELECT
    MAX(c.age) max_age
FROM
    hh.candidate c

1 rows affected.


[(100,)]

- <span style="color:blue">100 years is a rather mysterious age for employment. It needs to explore the sampling range in more detail.</span>

<span style="color:red">2.2 Calculation of the minimum age (min_age) of the candidate in the table.</span>

In [8]:
%%sql $conn
/* using MIN function to get the minimum val */ 
SELECT
    MIN(c.age) min_age
FROM
    hh.candidate c

1 rows affected.


[(14,)]

- <span style="color:blue">14 years old may seem too early here, but not the most anomalous either - the law allows you to start working a little at this age. Nothing unusual.</span>

<span style="color:red">2.3 Minor data cleanup. Writing a query that will allow to calculate for each age (age) how many (cnt) people of this age we have. Sorting the result by age in reverse order.</span>

In [9]:
%%sql $conn
/* using DISTINCT to get grouping by unique vals
and ordering output to see the largest ones */
SELECT
    DISTINCT c.age,
    COUNT(c.id) cnt
FROM
    hh.candidate c
GROUP BY 1
ORDER BY 1 DESC

63 rows affected.


[(100, 1),
 (77, 1),
 (76, 1),
 (73, 4),
 (72, 3),
 (71, 4),
 (70, 3),
 (69, 4),
 (68, 4),
 (67, 9),
 (66, 11),
 (65, 20),
 (64, 17),
 (63, 29),
 (62, 22),
 (61, 38),
 (60, 59),
 (59, 72),
 (58, 86),
 (57, 131),
 (56, 120),
 (55, 137),
 (54, 186),
 (53, 169),
 (52, 186),
 (51, 217),
 (50, 240),
 (49, 286),
 (48, 353),
 (47, 383),
 (46, 461),
 (45, 449),
 (44, 566),
 (43, 623),
 (42, 648),
 (41, 721),
 (40, 877),
 (39, 999),
 (38, 1110),
 (37, 1276),
 (36, 1387),
 (35, 1734),
 (34, 1836),
 (33, 2187),
 (32, 2531),
 (31, 2675),
 (30, 2834),
 (29, 2765),
 (28, 2748),
 (27, 2583),
 (26, 2372),
 (25, 2212),
 (24, 2074),
 (23, 1582),
 (22, 1284),
 (21, 754),
 (20, 405),
 (19, 173),
 (18, 61),
 (17, 14),
 (16, 4),
 (15, 2),
 (14, 1)]

- <span style="color:blue">There is only one outlier value that makes sense to exclude from the sample.</span>

<span style="color:red">2.4 According to Rosstat, the average age of those employed in the Russian economy is 39.7 years. Rounding this value up to 40, let's find the number of candidates who are older than this age. It' important do not forget to filter out the "erroneous" age of 100.</span>

In [10]:
%%sql $conn
/* filtering output of COUNT function by proper range to get the val */
SELECT
    COUNT(c.id) cnt
FROM
    hh.candidate c
WHERE c.age BETWEEN 41 AND 99

1 rows affected.


[(6263,)]

- <span style="color:blue">Well, let's do some math here, guys:</span>

In [11]:
prc = round(6263 / 44744 * 100)
display(f'The percentage of older job seekers is only {prc}% of the total amount')

'The percentage of older job seekers is only 14% of the total amount'

- <span style="color:blue">A simple calculation shows that we have the only 14% of older applicants (6263 out of almost 45000) here. Not so many! The vast majority of applicants are younger. One can raise the question of the advisability of the recent pension reform.</span>

### <div style="text-align:center"><span style="color:red">3. Global indicator analysis.</span></div>

<span style="color:red">3.1 To get started, let's write a query that will let us know how many (cnt) candidates we have from each city. Sample format: city, cnt. Let's group the result by the title column and sort by number in reverse order.</span>

In [12]:
%%sql $conn
/* joining the tables, grouping by city to get candidate quantity 
for each one and ordering output to see the highest vals */
SELECT
    ct.title city,
    COUNT(c.id) cnt
FROM
    hh.city ct
JOIN hh.candidate c ON ct.id = c.city_id
GROUP BY 1
ORDER BY 2 DESC

984 rows affected.


[('Москва', 16622),
 ('Санкт-Петербург', 4937),
 ('Краснодар', 1066),
 ('Новосибирск', 958),
 ('Казань', 872),
 ('Екатеринбург', 734),
 ('Самара', 703),
 ('Ростов-на-Дону', 607),
 ('Нижний Новгород', 598),
 ('Уфа', 565),
 ('Алматы', 556),
 ('Воронеж', 538),
 ('Пермь', 445),
 ('Красноярск', 407),
 ('Тюмень', 355),
 ('Минск', 353),
 ('Челябинск', 330),
 ('Омск', 303),
 ('Томск', 285),
 ('Ярославль', 282),
 ('Волгоград', 272),
 ('Саратов', 260),
 ('Калининград', 219),
 ('Иркутск', 213),
 ('Владивосток', 210),
 ('Тольятти', 198),
 ('Хабаровск', 198),
 ('Ижевск', 197),
 ('Сочи', 193),
 ('Барнаул', 191),
 ('Оренбург', 183),
 ('Ульяновск', 168),
 ('Липецк', 163),
 ('Тверь', 162),
 ('Ставрополь', 157),
 ('Кемерово', 156),
 ('Тула', 153),
 ('Набережные Челны', 152),
 ('Симферополь', 151),
 ('Пенза', 149),
 ('Калуга', 147),
 ('Нур-Султан', 143),
 ('Рязань', 139),
 ('Чебоксары', 134),
 ('Белгород', 126),
 ('Подольск (Московская область)', 125),
 ('Балашиха', 119),
 ('Иваново (Ивановская область)', 119),
 ('Курск', 119),
 ('Химки', 118),
 ('Зеленоград', 115),
 ('Брянск', 112),
 ('Астана', 111),
 ('Одинцово', 111),
 ('Мытищи', 111),
 ('Киров', 109),
 ('Астрахань', 105),
 ('Королев', 105),
 ('Сургут', 103),
 ('Вологда', 103),
 ('Севастополь', 101),
 ('Смоленск', 101),
 ('Владимир', 97),
 ('Орел', 95),
 ('Домодедово', 87),
 ('Тамбов', 86),
 ('Люберцы', 84),
 ('Новокузнецк', 80),
 ('Мурманск', 79),
 ('Саранск', 78),
 ('Раменское', 76),
 ('Красногорск', 75),
 ('Магнитогорск', 75),
 ('Йошкар-Ола', 75),
 ('Кострома', 72),
 ('Сыктывкар', 71),
 ('Петрозаводск', 70),
 ('Киев', 69),
 ('Архангельск', 66),
 ('Великий Новгород', 66),
 ('Щелково', 65),
 ('Новороссийск', 65),
 ('Улан-Удэ', 63),
 ('Чита', 62),
 ('Обнинск', 60),
 ('Пушкино (Московская область)', 60),
 ('Таганрог', 58),
 ('Караганда', 57),
 ('Нижневартовск', 54),
 ('Железнодорожный', 53),
 ('Лобня', 53),
 ('Череповец', 52),
 ('Стерлитамак', 48),
 ('Махачкала', 47),
 ('Актобе', 45),
 ('Жуковский', 44),
 ('Долгопрудный', 44),
 ('Благовещенск (Амурская область)', 43),
 ('Шымкент', 43),
 ('Видное', 43),
 ('Волжский', 42),
 ('Нижний Тагил', 41),
 ('Серпухов', 39),
 ('Чехов', 39),
 ('Сергиев Посад', 39),
 ('Абакан', 39),
 ('Псков', 39),
 ('Альметьевск', 39),
 ('Дзержинск (Нижегородская область)', 38),
 ('Пятигорск', 36),
 ('Дмитров', 35),
 ('Ханты-Мансийск', 34),
 ('Владикавказ', 34),
 ('Энгельс', 33),
 ('Рыбинск', 33),
 ('Старый Оскол', 33),
 ('Актау', 32),
 ('Атырау', 31),
 ('Комсомольск-на-Амуре', 29),
 ('Курган', 29),
 ('Южно-Сахалинск', 28),
 ('Бердск', 28),
 ('Электросталь', 28),
 ('Ангарск', 28),
 ('Ташкент', 27),
 ('Адлер', 27),
 ('Бийск', 27),
 ('Новочеркасск', 27),
 ('Ивантеевка', 27),
 ('Анапа', 26),
 ('Наро-Фоминск', 26),
 ('Нижнекамск', 26),
 ('Ноябрьск', 25),
 ('Якутск', 24),
 ('Петропавловск-Камчатский', 24),
 ('Салават', 24),
 ('Тобольск', 24),
 ('Елабуга', 23),
 ('Клин', 23),
 ('Миасс', 23),
 ('Уральск', 23),
 ('Братск', 23),
 ('Павлодар', 22),
 ('Шахты', 22),
 ('Коломна', 22),
 ('Зеленодольск (Республика Татарстан)', 21),
 ('Краснознаменск (Московская область)', 21),
 ('Ступино', 21),
 ('Норильск', 21),
 ('Усть-Каменогорск', 21),
 ('Солнечногорск', 21),
 ('Нефтекамск', 21),
 ('Гомель', 21),
 ('Нефтеюганск', 21),
 ('Новый Уренгой', 20),
 ('Истра', 20),
 ('Грозный', 19),
 ('Новочебоксарск', 19),
 ('Балаково', 19),
 ('Бишкек', 19),
 ('Ессентуки', 19),
 ('Ухта', 19),
 ('Бузулук', 19),
 ('Ногинск', 18),
 ('Брест', 18),
 ('Реутов', 18),
 ('Всеволожск', 18),
 ('Орск', 18),
 ('Новомосковск (Тульская область)', 18),
 ('Фрязино', 17),
 ('Волгодонск', 17),
 ('Гатчина', 16),
 ('Батайск', 16),
 ('Котельники', 16),
 ('Армавир', 16),
 ('Первоуральск', 16),
 ('Донецк (Украина)', 16),
 ('Ялта', 16),
 ('Нахабино', 15),
 ('Орехово-Зуево', 15),
 ('Сызрань', 15),
 ('Северск(Томская область)', 15),
 ('Голицыно', 15),
 ('Майкоп', 15),
 ('Тараз', 14),
 ('Кропоткин', 14),
 ('Златоуст', 14),
 ('Муром', 14),
 ('Семей', 14),
 ('Троицк (Московская область)', 14),
 ('Магадан', 13),
 ('Октябрьский (Республика

- <span style="color:blue">Moscow's huge lead compared to other cities is not due to anomalies in the mathematical distribution, but to the anomalous distribution of real resources in the economy and mediocre domestic politics. The fat capital and the dying depression regions are one of the grim signs of the current crazy times.</span>

In [13]:
# changing the rows display limitation:
%config SqlMagic.displaylimit=10

<span style="color:red">3.2 Anyway, Moscow is conspicuous as perhaps the most active labor market. Let's write a query that will allow to understand which candidates from Moscow will be satisfied with the “project work”. Sample format: gender, age, desirable_occupation, city, employment_type, sorting: by candidate id.</span>

In [14]:
%%sql $conn
/* joining the tables and filtering/ordering output */
SELECT
    c.gender,
    c.age,
    c.desirable_occupation,
    ct.title city,
    c.employment_type
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
WHERE ct.title = 'Москва'
AND c.employment_type ~ 'проектная работа'
ORDER BY c.id

2950 rows affected.


[('M', 38, 'Веб-разработчик (HTML / CSS / JS / PHP / базы данных; фреймворки, дизайн, интерфейсы, CMS)', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 31, 'Специалист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('F', 42, 'pre-sale инженер, pre-sale менеджер', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 49, 'Дежурный администратор', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 29, 'Главный инженер проекта', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 22, 'Программист С++', 'Москва', 'проектная работа, частичная занятость'),
 ('F', 29, 'Технический специалист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 32, 'IT Operations Coordinator', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 23, 'Инженер-связист,системный администратор', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 31, 'Менеджер', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 44, 'Директор по развитию бизнеса / Директор по продажам', 'Москва', 'проектная работа, полная занятость'),
 ('M', 25, 'Frontend-разработчик', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 40, 'Менеджер проектов', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 30, 'Руководитель IT-департамента', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('M', 30, 'Программист', 'Москва', 'частичная занятость, проектная работа'),
 ('M', 35, 'Ruby / Rails разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 36, 'SAP consultant', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 37, 'IT - 1С специалист администратор систем', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 57, 'Специалист по топливному инжинирингу и мониторингу транспорта', 'Москва', 'проектная работа'),
 ('M', 27, 'Ведущий инженер', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 24, 'Junior Javascript Developer', 'Москва', 'стажировка, проектная работа, полная занятость'),
 ('M', 25, 'SMM-менеджер', 'Москва', 'проектная работа, частичная занятость, полная занятость'),
 ('M', 25, 'Дизайнер интерфейсов, веб-дизайнер, UI/UX дизайнер, Product designer', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 34, 'Design Engineer, Research Engineer', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 45, 'Инженер электросвязи,Монтажник связи, Монтажник слаботочных систем, Тех. поддержка.', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 31, 'Системный администратор', 'Москва', 'частичная занятость, проектная работа'),
 ('M', 33, 'Системный администратор', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('M', 28, 'Инженер', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 46, 'Контент-менеджер', 'Москва', 'стажировка, частичная занятость, проектная работа'),
 ('M', 30, 'Продвижение сайта, SEO специалист', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('F', 21, 'Менеджер проекта', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 34, 'Системный администратор, помощник системного администратора, инженер технической поддержки', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 32, 'Монтажник', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 28, 'Программист микроконтроллеров', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('M', 36, 'Программист-разработчик', 'Москва', 'частичная з

<span style="color:red">3.3 There was too much data. Let's filter only the most popular IT professions - developer, analyst, programmer. These names can be written in both capital and small letters. Sorting: by candidate id.</span>

In [15]:
%%sql $conn
/* just adding another filter to the previous request */
SELECT
    c.gender,
    c.age,
    c.desirable_occupation,
    ct.title city,
    c.employment_type
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
WHERE ct.title = 'Москва'
AND c.employment_type ~ 'проектная работа'
AND c.desirable_occupation ~* 'разработчик|программист|аналитик' 
ORDER BY c.id

778 rows affected.


[('M', 38, 'Веб-разработчик (HTML / CSS / JS / PHP / базы данных; фреймворки, дизайн, интерфейсы, CMS)', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 22, 'Программист С++', 'Москва', 'проектная работа, частичная занятость'),
 ('M', 25, 'Frontend-разработчик', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 30, 'Программист', 'Москва', 'частичная занятость, проектная работа'),
 ('M', 35, 'Ruby / Rails разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 28, 'Программист микроконтроллеров', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('M', 36, 'Программист-разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 25, 'Аналитик', 'Москва', 'проектная работа, стажировка, частичная занятость, полная занятость'),
 ('M', 38, 'Инженер, программист C/C++, разработчик ПО', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('F', 54, 'Ведущий инженер программист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 41, 'Программист С++', 'Москва', 'проектная работа, частичная занятость, полная занятость'),
 ('M', 43, 'Программист C++, мультимедийные приложения', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 41, 'Ведущий разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 33, 'Программист C++', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 40, 'Программист-разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 22, 'IOS разработчик', 'Москва', 'проектная работа, частичная занятость, полная занятость'),
 ('F', 27, 'Аналитик данных', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 47, 'Инженер-программист АСУ ТП', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 37, 'Инженер-программист', 'Москва', 'проектная работа, стажировка, частичная занятость, полная занятость'),
 ('M', 25, 'Аналитик', 'Москва', 'проектная работа, полная занятость'),
 ('M', 45, 'Инженер- программист 1С', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 30, 'Программист', 'Москва', 'волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 33, 'Аналитик отдела продаж', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 23, 'PHP-программист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 26, 'Project manager/Бизнес аналитик', 'Москва', 'проектная работа, полная занятость'),
 ('M', 37, 'Программист-разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('F', 54, 'Ведущий инженер программист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 24, 'PHP-программист', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('M', 34, 'Программист Си (middle) / Perl (junior)', 'Москва', 'стажировка, частичная занятость, проектная работа, полная занятость'),
 ('F', 30, 'Программист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 29, 'Web разработчик PHP, JS, MySQL, Linux (Senior)', 'Москва', 'проектная работа, частичная занятость, полная занятость'),
 ('M', 32, 'Инженер-программист (Software Engineer, CTO) .NET C#', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 58, 'разработчик Android/Linux', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 27, 'Frontend-разработчик', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 39, 'Программист', 'Москва', 'частичная занятость, проектная работа, полная занятость'),
 ('M', 24, 'Техник-программист', 'Москва', 'стажировка, волонтерство, частичная занятость, проектная работа, полная занятость'),
 ('M', 38, 'Прогр

- <span style="color:blue">As many as 778 qualified IT candidates right now - there is plenty to choose from. Looks nice! It remains to be hoped that at least one of them wrote the truth about himself.</span>

In [16]:
# changing the rows display limitation:
%config SqlMagic.displaylimit=5

<span style="color:red">3.4 For general information, let's try to select the numbers and cities of candidates whose position matches the desired one. Sample format: id, city with sorting the result by city and candidate id.</span>

In [17]:
%%sql $conn
/* joining the tables and filtering/ordering output */
SELECT
    c.id,
    ct.title city
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
WHERE c.desirable_occupation = c.current_occupation
ORDER BY 2, 1

5104 rows affected.


[(2009, 'Абакан'),
 (10340, 'Абакан'),
 (14449, 'Абакан'),
 (20261, 'Абакан'),
 (13705, 'Агрыз'),
 (967, 'Адлер'),
 (4276, 'Адлер'),
 (26878, 'Адлер'),
 (27717, 'Адлер'),
 (28057, 'Адлер'),
 (20058, 'Актау'),
 (25634, 'Актау'),
 (28152, 'Актау'),
 (30401, 'Актау'),
 (3631, 'Актобе'),
 (6103, 'Актобе'),
 (10471, 'Актобе'),
 (14077, 'Актобе'),
 (28675, 'Александров'),
 (81, 'Алматы'),
 (3453, 'Алматы'),
 (5118, 'Алматы'),
 (6524, 'Алматы'),
 (7267, 'Алматы'),
 (8253, 'Алматы'),
 (8882, 'Алматы'),
 (9369, 'Алматы'),
 (9902, 'Алматы'),
 (12208, 'Алматы'),
 (12783, 'Алматы'),
 (12834, 'Алматы'),
 (12953, 'Алматы'),
 (13986, 'Алматы'),
 (14211, 'Алматы'),
 (14696, 'Алматы'),
 (15001, 'Алматы'),
 (15224, 'Алматы'),
 (15614, 'Алматы'),
 (15615, 'Алматы'),
 (16317, 'Алматы'),
 (16615, 'Алматы'),
 (16759, 'Алматы'),
 (16763, 'Алматы'),
 (16893, 'Алматы'),
 (19438, 'Алматы'),
 (19545, 'Алматы'),
 (20284, 'Алматы'),
 (20730, 'Алматы'),
 (21038, 'Алматы'),
 (22357, 'Алматы'),
 (23461, 'Алматы'),
 (23579, 'Алматы'),
 (23989, 'Алматы'),
 (24257, 'Алматы'),
 (25273, 'Алматы'),
 (25401, 'Алматы'),
 (25441, 'Алматы'),
 (26564, 'Алматы'),
 (29737, 'Алматы'),
 (29835, 'Алматы'),
 (31490, 'Алматы'),
 (34298, 'Алматы'),
 (34408, 'Алматы'),
 (34499, 'Алматы'),
 (35431, 'Алматы'),
 (36239, 'Алматы'),
 (36257, 'Алматы'),
 (36809, 'Алматы'),
 (38845, 'Алматы'),
 (39131, 'Алматы'),
 (39411, 'Алматы'),
 (39848, 'Алматы'),
 (39981, 'Алматы'),
 (40401, 'Алматы'),
 (40750, 'Алматы'),
 (41192, 'Алматы'),
 (41936, 'Алматы'),
 (42233, 'Алматы'),
 (42670, 'Алматы'),
 (42964, 'Алматы'),
 (43560, 'Алматы'),
 (43879, 'Алматы'),
 (3763, 'Альметьевск'),
 (7301, 'Альметьевск'),
 (10869, 'Альметьевск'),
 (16683, 'Альметьевск'),
 (24801, 'Альметьевск'),
 (36915, 'Альметьевск'),
 (39968, 'Альметьевск'),
 (41295, 'Альметьевск'),
 (5032, 'Анапа'),
 (26758, 'Анапа'),
 (33712, 'Анапа'),
 (4627, 'Ангарск'),
 (4687, 'Ангарск'),
 (14460, 'Ангарск'),
 (17442, 'Ангарск'),
 (20372, 'Ангарск'),
 (22853, 'Ангарск'),
 (26430, 'Анжеро-Судженск'),
 (41201, 'Апрелевка'),
 (24290, 'Ардатов'),
 (446, 'Армавир'),
 (24827, 'Армения'),
 (33145, 'Артем'),
 (1979, 'Архангельск'),
 (7014, 'Архангельск'),
 (9116, 'Архангельск'),
 (19539, 'Архангельск'),
 (21061, 'Архангельск'),
 (21454, 'Архангельск'),
 (24674, 'Архангельск'),
 (29542, 'Архангельск'),
 (32032, 'Архангельск'),
 (40179, 'Архангельск'),
 (44434, 'Архангельск'),
 (800, 'Астана'),
 (7324, 'Астана'),
 (11162, 'Астана'),
 (12978, 'Астана'),
 (20573, 'Астана'),
 (26824, 'Астана'),
 (28459, 'Астана'),
 (29792, 'Астана'),
 (34557, 'Астана'),
 (34641, 'Астана'),
 (38817, 'Астана'),
 (40061, 'Астана'),
 (40435, 'Астана'),
 (41754, 'Астана'),
 (42488, 'Астана'),
 (42756, 'Астана'),
 (853, 'Астрахань'),
 (7339, 'Астрахань'),
 (7628, 'Астрахань'),
 (11478, 'Астрахань'),
 (12210, 'Астрахань'),
 (15449, 'Астрахань'),
 (15641, 'Астрахань'),
 (23230, 'Астрахань'),
 (25836, 'Астрахань'),
 (32496, 'Астрахань'),
 (33843, 'Астрахань'),
 (37381, 'Астрахань'),
 (37600, 'Астрахань'),
 (38997, 'Астрахань'),
 (3139, 'Атырау'),
 (13915, 'Атырау'),
 (15445, 'Атырау'),
 (21417, 'Атырау'),
 (31306, 'Атырау'),
 (15334, 'Ачинск'),
 (38855, 'Ачинск'),
 (9983, 'Баку'),
 (1537, 'Балаково'),
 (6026, 'Балаково'),
 (21395, 'Балаково'),
 (25230, 'Балаково'),
 (707, 'Балашиха'),
 (2951, 'Балашиха'),
 (3895, 'Балашиха'),
 (4860, 'Балашиха'),
 (10528, 'Балашиха'),
 (11239, 'Балашиха'),
 (15722, 'Балашиха'),
 (22226, 'Балашиха'),
 (23442, 'Балашиха'),
 (28061, 'Балашиха'),
 (30748, 'Балашиха'),
 (32282, 'Балашиха'),
 (33655, 'Балашиха'),
 (12552, 'Балхаш'),
 (4543, 'Барнаул'),
 (6082, 'Барнаул'),
 (7474, 'Барнаул'),
 (13917, 'Барнаул'),
 (14032, 'Барнаул'),
 (14252, 'Барнаул'),
 (18583, 'Барнаул'),
 (19859, 'Барнаул'),
 (24136, 'Барнаул'),
 (26077, 'Барнаул'),
 (26351, 'Барнаул'),
 (27768, 'Барнаул'),
 (29110, 'Барнаул'),
 (31435, 'Барнаул'),
 (35537, 'Барнаул'),
 (37401, 'Барнаул'),
 (37842, 'Барнаул'),
 (38417, 'Барнаул'),
 (5180, 'Батайск'),
 (6220, 'Батайск'),
 (8967,

- <span style="color:blue">Approximately one in ten applicants clearly indicate that they want to change jobs within their own competence. With the rest, things may be different, so this group, perhaps, looks pretty reliable. This criterion can be a real godsend for those hiring specialists who are not very inclined to delve into their business, but come to work mainly to shift papers.</span>

<span style="color:red">3.5 Determine the number of candidates of retirement age. The retirement age for men is 65 years, for women - 60 years.</span>

In [18]:
%%sql $conn
/* filtering output of COUNT() function by proper ranges to get the val */
SELECT
    COUNT(c.id)
FROM
    hh.candidate c
WHERE (c.gender = 'M' AND c.age BETWEEN 65 AND 99)
OR (c.gender = 'F' AND c.age BETWEEN 60 AND 99)

1 rows affected.


[(75,)]

- <span style="color:blue">Let's go back one more time to our mathematics from section 2.4. There we calculated the percentage of elderly applicants, and we got 14%. Let's do the same for the category of pensioners! And...</span>

In [19]:
prc = round(75 / 44744 * 100, 2)
display(f'The percentage of retirement age job seekers is only {prc}% of the total amount')

'The percentage of retirement age job seekers is only 0.17% of the total amount'

- <span style="color:blue">And it's 0.17%, guys! The only 0.17%! That's it. No more words about that pension reform...</span>

### <div style="text-align:center"><span style="color:red">4. Analysis of candidates for customers.</span></div>

In [20]:
# changing the rows display limitation:
%config SqlMagic.displaylimit=15

<span style="color:red">4.1 For a mining company, we need to select candidates from Novosibirsk, Omsk, Tomsk and Tyumen who are ready to work on a rotational basis. Sample format: gender, age, desirable_occupation, city, employment_type, timetable_type. Sort the result by city and candidate number.</span>

In [21]:
%%sql $conn
/* joining all the tables and filtering/ordering output */
SELECT
    c.gender,
    c.age,
    c.desirable_occupation,
    ct.title city,
    c.employment_type,
    t.title timetable_type
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
JOIN hh.candidate_timetable_type ctt ON ctt.candidate_id = c.id
JOIN hh.timetable_type t ON ctt.timetable_id = t.id
WHERE
    ct.title ~ 'Новосибирск|Омск|Томск|Тюмень'
AND t.title ~ 'вахтовый метод'
ORDER BY ct.title, c.id

11 rows affected.


[('M', 29, 'ИТ Инженер', 'Новосибирск', 'полная занятость', 'вахтовый метод'),
 ('M', 25, 'Заместитель начальника лаборатории', 'Новосибирск', 'проектная работа, стажировка, частичная занятость, полная занятость', 'вахтовый метод'),
 ('M', 30, 'Ведущий инженер, Специалист по защите информации,', 'Новосибирск', 'частичная занятость, полная занятость', 'вахтовый метод'),
 ('M', 23, 'Программист', 'Новосибирск', 'полная занятость', 'вахтовый метод'),
 ('M', 35, 'Инженер АСУТП, инженер-электроник', 'Омск', 'полная занятость', 'вахтовый метод'),
 ('M', 25, 'Тестировщик ПО', 'Омск', 'стажировка, полная занятость', 'вахтовый метод'),
 ('M', 26, 'Специалист технической поддержки', 'Томск', 'частичная занятость, полная занятость', 'вахтовый метод'),
 ('M', 30, 'Менеджер проектов', 'Томск', 'проектная работа, частичная занятость, полная занятость', 'вахтовый метод'),
 ('M', 42, 'Инженер', 'Томск', 'проектная работа, частичная занятость, полная занятость', 'вахтовый метод'),
 ('M', 31, 'Инженер связи', 'Тюмень', 'полная занятость', 'вахтовый метод'),
 ('M', 31, 'Инженер АСУ ТП, АСУ, Мастер КИП, Программист АСУ', 'Тюмень', 'полная занятость', 'вахтовый метод')]

- <span style="color:blue">Not so much! Well, it was not so simply that it has become now to find real romantic heroes who could dig the whole earth in search of genuine treasures ...</span>

<span style="color:red">4.2 For customers from St. Petersburg, we need to collect a list of 10 desired professions of candidates from the same city from 16 to 21 years old (the sample includes 16 and 21, sorted by age) indicating their age, and also add the Total line with the total number such candidates. Let's write a query that will allow you to get a selection.</span>

In [22]:
%%sql $conn
/* union (all) of two joined/filtered/ordered tables output with
additional COUNT() function output joined/filtered the same way */
(SELECT
    c.desirable_occupation,
    c.age
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
WHERE
    c.age BETWEEN 16 AND 21
AND ct.title = 'Санкт-Петербург'
ORDER BY 2
LIMIT 10)
UNION ALL
SELECT
    'Total',
    COUNT(c.id)
FROM
    hh.candidate c
JOIN hh.city ct ON ct.id = c.city_id
WHERE
    c.age BETWEEN 16 AND 21
AND ct.title = 'Санкт-Петербург'

11 rows affected.


[('Системный администратор', 16),
 ('Junior Разработчик C++/C#', 18),
 ('Программист', 18),
 ('Junior Data Scientist', 18),
 ('Руководитель web-разработки', 18),
 ('Специалист по IT', 18),
 ('Unity3D developer Junior/middle', 18),
 ('HTML-верстальщик', 18),
 ('3D-дизайнер', 18),
 ('Java-разработчик', 18),
 ('Total', 161)]

- <span style="color:blue">... But there are more than enough lovers to sit all their lives in offices and at home! The abundance of vacancies in St. Petersburg is easily explained by the fact that, St. Petersburg is beautiful and will become even more beautiful, unlike some other "capitals"!</span>

### <div style="text-align:center"><span style="color:red">5. The Final Conclusion:</span></div>

### <div style="text-align:center"><span style="color:blue">Oh, those Russians...</span></div>